<a href="https://colab.research.google.com/github/Viviankatheu/Electric-Car---Sharing/blob/main/Moringa_Data_Science_Prep_W4_Independent_Project_2019_07_Vivian_Katheu_Colab_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Electric Car Sharing Project - Week 4 IP** 

##1.0. Importing our Libraries

In [1]:
# We shall import the pandas and numpy libraries

# Importing the pandas library
import pandas as pd

# Importing the numpy library
import numpy as np


##1.1 Reading the dataset from our CSV File

Dataset link for download or access: http://bit.ly/autolib_dataset

In [2]:
# Let's read the data from the CSV file and create the dataframe to be used

df = pd.read_csv('Autolib_dataset (2).csv')
df

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]

##1.2 Previewing our Dataset

In [3]:
# Let's preview the first 10 rows of our data
# 
df.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

##1.3 Accessing Information about our Dataset

In [4]:
#Now to access information from the table

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

##1.4 Cleaning our Dataset

Let us perform data cleaning procedures below providing a documentation for our actions and reasons. We will perform as many data cleaning procedures as we think suitable for the various dimensions of data.

### *1.41 ) Validity: Challenges*

In [5]:
# Procedure 1: Irrelevant Data Observation
# Data Cleaning Action: Dropping Displayed Comment Column
# Explanation: We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Displayed comment', axis =1, inplace=True )
df

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 24 columns]

In [6]:
# Procedure 2: Irrelevant Data Observation
# Data Cleaning Action: Dropping Scheduled at Column
# Explanation: We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Scheduled at', axis =1, inplace=True )
df

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 23 columns]

In [7]:
# Procedure 3: Irrelevant Data Observation
# Data Cleaning Action: Dropping Cars Column
# Explanation: We won't need it during Analysis. We've been told its redundant since its the same as the Blue cars Column
#
df.drop(columns= 'Cars', axis =1, inplace=True )
df

Address  Bluecar counter  ...  hour  minute
0                    2 Avenue de Suffren                0  ...    11      43
1              145 Rue Raymond Losserand                6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3                 51 Rue Eugène Oudiné                3  ...     4      37
4     6 avenue de la Porte de Champerret                3  ...    17      23
...                                  ...              ...  ...   ...     ...
4995                 8 avenue Ménelotte                2  ...    11      26
4996                   37 rue de Dantzig                4  ...    16      56
4997                      142 rue du Bac                1  ...     7       1
4998         2 avenue du Val de Fontenay                2  ...    17      27
4999                 5 Rue Henry Brisson                4  ...    10       2

[5000 rows x 22 columns]

In [8]:
# Procedure 4: Irrelevant Data Observation
# Data Cleaning Action: Dropping Charge Slots Column
# Explanation:  We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Charge Slots', axis =1, inplace=True )
df


Address  Bluecar counter  ...  hour  minute
0                    2 Avenue de Suffren                0  ...    11      43
1              145 Rue Raymond Losserand                6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3                 51 Rue Eugène Oudiné                3  ...     4      37
4     6 avenue de la Porte de Champerret                3  ...    17      23
...                                  ...              ...  ...   ...     ...
4995                 8 avenue Ménelotte                2  ...    11      26
4996                   37 rue de Dantzig                4  ...    16      56
4997                      142 rue du Bac                1  ...     7       1
4998         2 avenue du Val de Fontenay                2  ...    17      27
4999                 5 Rue Henry Brisson                4  ...    10       2

[5000 rows x 21 columns]

In [9]:
# Procedure 4: Irrelevant Data Observation
# Data Cleaning Action: Dropping Charging Status Column
# Explanation:  We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Charging Status', axis =1, inplace=True )
df

Address  Bluecar counter  ...  hour  minute
0                    2 Avenue de Suffren                0  ...    11      43
1              145 Rue Raymond Losserand                6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3                 51 Rue Eugène Oudiné                3  ...     4      37
4     6 avenue de la Porte de Champerret                3  ...    17      23
...                                  ...              ...  ...   ...     ...
4995                 8 avenue Ménelotte                2  ...    11      26
4996                   37 rue de Dantzig                4  ...    16      56
4997                      142 rue du Bac                1  ...     7       1
4998         2 avenue du Val de Fontenay                2  ...    17      27
4999                 5 Rue Henry Brisson                4  ...    10       2

[5000 rows x 20 columns]

In [10]:
# Procedure 4: Irrelevant Data Observation
# Data Cleaning Action: Dropping Public Name Column
# Explanation: This column and the ID column are the exact same therefore we drop this column 
#
df.drop(columns= 'Public name', axis =1, inplace=True )
df


Address  Bluecar counter  ...  hour  minute
0                    2 Avenue de Suffren                0  ...    11      43
1              145 Rue Raymond Losserand                6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3                 51 Rue Eugène Oudiné                3  ...     4      37
4     6 avenue de la Porte de Champerret                3  ...    17      23
...                                  ...              ...  ...   ...     ...
4995                 8 avenue Ménelotte                2  ...    11      26
4996                   37 rue de Dantzig                4  ...    16      56
4997                      142 rue du Bac                1  ...     7       1
4998         2 avenue du Val de Fontenay                2  ...    17      27
4999                 5 Rue Henry Brisson                4  ...    10       2

[5000 rows x 19 columns]

In [11]:
# Procedure 5: Irrelevant Data Observation
# Data Cleaning Action: Dropping Geo point Column
# Explanation: We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Geo point', axis =1, inplace=True )
df.head()


Address  Bluecar counter  ...  hour  minute
0                 2 Avenue de Suffren                0  ...    11      43
1           145 Rue Raymond Losserand                6  ...     7      24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3              51 Rue Eugène Oudiné                3  ...     4      37
4  6 avenue de la Porte de Champerret                3  ...    17      23

[5 rows x 18 columns]

In [12]:
# Procedure 8: Irrelevant Data Observation
# Data Cleaning Action: Dropping Subscription status Column
# Explanation: We won't need it during Analysis. No question to be answered requires that column.
#
df.drop(columns= 'Subscription status', axis =1, inplace=True )
df.head()

Address  Bluecar counter  ...  hour  minute
0                 2 Avenue de Suffren                0  ...    11      43
1           145 Rue Raymond Losserand                6  ...     7      24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3              51 Rue Eugène Oudiné                3  ...     4      37
4  6 avenue de la Porte de Champerret                3  ...    17      23

[5 rows x 17 columns]

In [13]:
# Procedure 9: Irrelevant Data Observation
# Data Cleaning Action: Dropping the Year and Month Columns
# Explanation: We already know that the data was collected in the month of April, 2018

df.drop(columns='year', axis =1, inplace=True )
df.head()


Address  Bluecar counter  ...  hour  minute
0                 2 Avenue de Suffren                0  ...    11      43
1           145 Rue Raymond Losserand                6  ...     7      24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3              51 Rue Eugène Oudiné                3  ...     4      37
4  6 avenue de la Porte de Champerret                3  ...    17      23

[5 rows x 16 columns]

In [14]:
# Procedure 10: Irrelevant Data Observation
# Data Cleaning Action: Dropping the Year and Month Columns
# Explanation: We already know that the data was collected in the month of April, 2018

df.drop(columns='month', axis =1, inplace=True )
df.head()

Address  Bluecar counter  ...  hour  minute
0                 2 Avenue de Suffren                0  ...    11      43
1           145 Rue Raymond Losserand                6  ...     7      24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3              51 Rue Eugène Oudiné                3  ...     4      37
4  6 avenue de la Porte de Champerret                3  ...    17      23

[5 rows x 15 columns]

###*1.42 ) Accuracy*


In [15]:
# Data Cleaning Action: No action taken
# Explanation: The data provided is accurate

###*1.43 ) Uniformity*

In [16]:
# Data Cleaning Action: Renaming column names to upper case
# Explanation: To make the table look more presentable

df.columns = ['ADDRESS','BLUECAR COUNTER', 'UTILIB COUNTER','UTILIB 1.4 COUNTER', 'CITY', 'ID', 'KIND', 'POSTAL CODE', 'RENTAL STATUS', 'SLOTS', 'STATION TYPE', 'STATUS', 'DAY', 'HOUR', 'MINUTE']
df

ADDRESS  BLUECAR COUNTER  ...  HOUR  MINUTE
0                    2 Avenue de Suffren                0  ...    11      43
1              145 Rue Raymond Losserand                6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy                3  ...    20      14
3                 51 Rue Eugène Oudiné                3  ...     4      37
4     6 avenue de la Porte de Champerret                3  ...    17      23
...                                  ...              ...  ...   ...     ...
4995                 8 avenue Ménelotte                2  ...    11      26
4996                   37 rue de Dantzig                4  ...    16      56
4997                      142 rue du Bac                1  ...     7       1
4998         2 avenue du Val de Fontenay                2  ...    17      27
4999                 5 Rue Henry Brisson                4  ...    10       2

[5000 rows x 15 columns]

###*1.44 ) Completeness*

In [17]:
# Procedure 1:
# Data Cleaning Action:
# Explanation: lets check whether there are any missing valus
#
df.isnull().sum()

ADDRESS               0
BLUECAR COUNTER       0
UTILIB COUNTER        0
UTILIB 1.4 COUNTER    0
CITY                  0
ID                    0
KIND                  0
POSTAL CODE           0
RENTAL STATUS         0
SLOTS                 0
STATION TYPE          0
STATUS                0
DAY                   0
HOUR                  0
MINUTE                0
dtype: int64

There are no missing values

###*1.45 ) Consistency*

In [18]:
# Procedure 1: Duplicates
# Data Cleaning Action: No action taken
# Explanation: There were no duplicate instances

df.duplicated().value_counts()

False    5000
dtype: int64

There are no duplicates

In [19]:
df.shape

(5000, 15)

##1.5 Exporting the Cleaned Dataset

In [21]:
# Let's export our dataframe into a csv file as shown 
# we will use the to_csv function to create a csv file with the name Clean_Autolib
# and export it

df.to_csv('Clean_Autolib.csv')

##1.6 Analyzing the Data

In [22]:
# We shall use the cleaned data we've just exported

df1 = pd.read_csv('Clean_Autolib.csv')
df1.head()

Unnamed: 0                             ADDRESS  ...  HOUR  MINUTE
0           0                 2 Avenue de Suffren  ...    11      43
1           1           145 Rue Raymond Losserand  ...     7      24
2           2    2 Avenue John Fitzgerald Kennedy  ...    20      14
3           3              51 Rue Eugène Oudiné  ...     4      37
4           4  6 avenue de la Porte de Champerret  ...    17      23

[5 rows x 16 columns]

### 1.61 ) Finding out the most popular hour for picking up a shared electric car In Paris in April, 2018

In [25]:
hr = df1[df1['CITY']== 'Paris']
pophr = hr.groupby(['ID', 'HOUR'])[['BLUECAR COUNTER']].sum()
pophr.head()

BLUECAR COUNTER
ID                 HOUR                 
paris-25aout1944-8 3                   3
                   5                   0
                   7                   4
                   12                  2
                   18                  2

In [26]:
pophr['BLUECAR_USAGE'] = pophr['BLUECAR COUNTER'].diff()
pophr.head()

BLUECAR COUNTER  BLUECAR_USAGE
ID                 HOUR                                
paris-25aout1944-8 3                   3            NaN
                   5                   0           -3.0
                   7                   4            4.0
                   12                  2           -2.0
                   18                  2            0.0

In [27]:
pop = pophr[pophr['BLUECAR_USAGE'] < 0]
pop.head()

BLUECAR COUNTER  BLUECAR_USAGE
ID                  HOUR                                
paris-25aout1944-8  5                   0           -3.0
                    12                  2           -2.0
                    20                  1           -1.0
paris-adolphemax-6  13                  0           -4.0
paris-adolpheyvon-6 6                   3           -1.0

We will create a new column to represent the difference in cars at the station

In [33]:
#Let's find the difference using the diff method

df1['BLUECAR_USAGE'] = df1['BLUECAR COUNTER'].diff()

#create a column for utilib usage and another for utilib 1.4 usage
df1['UTILIB_USAGE'] = df1['UTILIB COUNTER'].diff()
df1['UTILIB_1.4_USAGE'] = df1['UTILIB 1.4 COUNTER'].diff()

df1.head()

Unnamed: 0  ... UTILIB_1.4_USAGE
0           0  ...              NaN
1           1  ...              0.0
2           2  ...              2.0
3           3  ...             -2.0
4           4  ...              0.0

[5 rows x 19 columns]

In [39]:
#looking for missing values
df1.isnull().sum()

Unnamed: 0            0
ADDRESS               0
BLUECAR COUNTER       0
UTILIB COUNTER        0
UTILIB 1.4 COUNTER    0
CITY                  0
ID                    0
KIND                  0
POSTAL CODE           0
RENTAL STATUS         0
SLOTS                 0
STATION TYPE          0
STATUS                0
DAY                   0
HOUR                  0
MINUTE                0
BLUECAR_USAGE         1
UTILIB_USAGE          1
UTILIB_1.4_USAGE      1
dtype: int64

In [40]:
#We shall fill the missing values with 0

df2 = df1.fillna(0)
df2.head()

Unnamed: 0  ... UTILIB_1.4_USAGE
0           0  ...              0.0
1           1  ...              0.0
2           2  ...              2.0
3           3  ...             -2.0
4           4  ...              0.0

[5 rows x 19 columns]

In [58]:
df2

Unnamed: 0  ... UTILIB_1.4_USAGE
0              0  ...              0.0
1              1  ...              0.0
2              2  ...              2.0
3              3  ...             -2.0
4              4  ...              0.0
...          ...  ...              ...
4995        4995  ...              0.0
4996        4996  ...              0.0
4997        4997  ...              0.0
4998        4998  ...              0.0
4999        4999  ...              0.0

[5000 rows x 19 columns]

#### 1.61 a) Most popular hour for picking up a blue car

In [44]:
#Let's find out the most popular hour for picking up a blue car
#Note that the usage column is negative when a car has been picked

df = df2[(df2['CITY'] == 'Paris') & (df1['BLUECAR_USAGE'] < 0)]
 
bluecar = df.groupby(['HOUR'])[['BLUECAR_USAGE']].sum()
bluecar.head()

BLUECAR_USAGE
HOUR               
0            -146.0
1            -124.0
2            -150.0
3            -164.0
4            -160.0

In [49]:
bluecar1 = bluecar.sort_values('BLUECAR_USAGE', ascending = True)
bluecar1

BLUECAR_USAGE
HOUR               
21           -231.0
7            -196.0
20           -193.0
5            -190.0
16           -172.0
23           -165.0
3            -164.0
13           -160.0
4            -160.0
12           -159.0
19           -154.0
22           -153.0
8            -153.0
17           -152.0
2            -150.0
9            -149.0
18           -147.0
0            -146.0
10           -146.0
6            -140.0
11           -136.0
15           -125.0
1            -124.0
14           -123.0

Therefore, the most popular hour for picking up a blue car is 2100hrs. 231 cars are picked at that time.

#### 1.61 b) Most popular hour for picking up a utilib

In [51]:
ut= df.groupby('HOUR')[['UTILIB_USAGE']].sum()
ut.head()

UTILIB_USAGE
HOUR              
0             -3.0
1             -2.0
2             -2.0
3             -5.0
4             -2.0

In [52]:
#We will now sort the values in ascending order
utilib = ut.sort_values('UTILIB_USAGE', ascending = True)
utilib.head()

UTILIB_USAGE
HOUR              
19            -8.0
22            -8.0
14            -6.0
23            -5.0
3             -5.0

Most utilibs are picked up at 1900 and 2200 hours meaning these are the most popular hours for picking utilibs. 8 utilibs are picked up at this time

####1.61 c) Most popular hour for picking up utilib 1.4 

In [53]:
u = df.groupby('HOUR')[['UTILIB_1.4_USAGE']].sum()
u.head()

UTILIB_1.4_USAGE
HOUR                  
0                -13.0
1                  2.0
2                 -3.0
3                 -8.0
4                 -3.0

In [54]:
#sorting these values
uti = u.sort_values('UTILIB_1.4_USAGE', ascending = True)
uti

UTILIB_1.4_USAGE
HOUR                  
0                -13.0
8                 -9.0
3                 -8.0
13                -6.0
19                -5.0
7                 -5.0
10                -4.0
12                -4.0
2                 -3.0
4                 -3.0
16                -3.0
20                -2.0
17                -2.0
11                -2.0
14                -1.0
18                -1.0
5                 -1.0
23                -1.0
6                  0.0
22                 1.0
9                  2.0
1                  2.0
21                 2.0
15                 4.0

Therefore, the most popular hour for picking up a utilib 1.4 is 0000 hrs. 13 of them are picked up at this time.

### 1.62) Finding out the most popular hour for returning cars in Paris in April, 2018

#### 1.62 a) Most popular hour for returning a blue car

In [61]:
#Remember that now the opposite is true. A column has a positive value if a car has been returned

rb = df2[(df2['CITY'] == 'Paris') & (df2['BLUECAR_USAGE'] > 0)]
rb.head()

Unnamed: 0  ... UTILIB_1.4_USAGE
1            1  ...              0.0
9            9  ...              0.0
12          12  ...              0.0
17          17  ...              0.0
26          26  ...              1.0

[5 rows x 19 columns]

In [62]:
return_b = rb.groupby(['HOUR'])[['BLUECAR_USAGE']].sum()
return_b.head()

BLUECAR_USAGE
HOUR               
0             100.0
1             106.0
2             106.0
3             134.0
4              84.0

In [64]:
#Let's sort this data in ascending order

return_blue = return_b.sort_values('BLUECAR_USAGE', ascending= False)
return_blue.head()

BLUECAR_USAGE
HOUR               
3             134.0
12            124.0
6             118.0
10            112.0
8             112.0

We find out that the most popular hour for returning a blue car is 0300 hrs


#### 1.62 b) Most popular hour for returning a utilib





In [66]:

ru = df2[(df2['CITY'] == 'Paris') & (df2['UTILIB_USAGE'] > 0)]
ru.head()

Unnamed: 0  ... UTILIB_1.4_USAGE
3             3  ...             -2.0
12           12  ...              0.0
72           72  ...              0.0
77           77  ...              0.0
129         129  ...             -1.0

[5 rows x 19 columns]

In [70]:
return_u = ru.groupby(['HOUR'])[['UTILIB_USAGE']].sum()
return_u.head()

UTILIB_USAGE
HOUR              
0              5.0
1              3.0
2              6.0
3              5.0
4              3.0

In [72]:
return_utilib = return_u.sort_values('UTILIB_USAGE', ascending= False)
return_utilib.head()

UTILIB_USAGE
HOUR              
6             12.0
8             12.0
17             9.0
11             9.0
21             8.0

The most popular hour for returning a utilib is 0600 and 0800 hours. 12 utilibs are returned at that time

#### 1.62 c) Most popular hour for returning a utilib 1.4

In [73]:
rut = df2[(df2['CITY'] == 'Paris') & (df2['UTILIB_1.4_USAGE'] > 0)]
rut.head()

Unnamed: 0  ... UTILIB_1.4_USAGE
19          19  ...              2.0
26          26  ...              1.0
49          49  ...              1.0
62          62  ...              2.0
67          67  ...              1.0

[5 rows x 19 columns]

In [75]:
return_ut = rut.groupby(['HOUR'])[['UTILIB_1.4_USAGE']].sum()
return_ut.head()

UTILIB_1.4_USAGE
HOUR                  
0                  8.0
1                 11.0
2                  8.0
3                 13.0
4                  8.0

In [76]:
return_r = return_ut.sort_values('UTILIB_1.4_USAGE', ascending= False)
return_r.head()

UTILIB_1.4_USAGE
HOUR                  
9                 18.0
17                18.0
5                 17.0
21                16.0
10                15.0

The most popular hour for returning a utilib 1.4 is 0900 and 1700 hours

### 1.63) Finding out the most popular station overall and at the most picking hour

#### 1.63 a) Finding out the most popular station overall

In [77]:
df2

Unnamed: 0  ... UTILIB_1.4_USAGE
0              0  ...              0.0
1              1  ...              0.0
2              2  ...              2.0
3              3  ...             -2.0
4              4  ...              0.0
...          ...  ...              ...
4995        4995  ...              0.0
4996        4996  ...              0.0
4997        4997  ...              0.0
4998        4998  ...              0.0
4999        4999  ...              0.0

[5000 rows x 19 columns]

In [78]:
station2 = df2[(df2['STATUS'] == 'ok') & (df2['KIND'] == 'STATION')]
station2.head()

Unnamed: 0                           ADDRESS  ...  UTILIB_USAGE  UTILIB_1.4_USAGE
0           0               2 Avenue de Suffren  ...           0.0               0.0
1           1         145 Rue Raymond Losserand  ...           0.0               0.0
2           2  2 Avenue John Fitzgerald Kennedy  ...           0.0               2.0
3           3            51 Rue Eugène Oudiné  ...           1.0              -2.0
5           5              8 Boulevard Voltaire  ...           0.0               0.0

[5 rows x 19 columns]

In [79]:
pops = station2.groupby('ID')[['ID']].count()
pops.head()

ID
ID                                  
alfortville-charlesdegaulle-16     5
alfortville-etiennedolet-174       2
alfortville-jeanbaptistepreux-17   4
alfortville-josephfranceschi-1     5
alfortville-portalanglais-39       7

In [81]:
pops.columns = ['ID COUNT']
pops.head()

ID COUNT
ID                                        
alfortville-charlesdegaulle-16           5
alfortville-etiennedolet-174             2
alfortville-jeanbaptistepreux-17         4
alfortville-josephfranceschi-1           5
alfortville-portalanglais-39             7

In [83]:
popstation = pops.sort_values('ID COUNT', ascending= False)
popstation.head()

ID COUNT
ID                                
paris-portedemontrouge-8        13
paris-voltaire-182              11
paris-mathis-35                 11
paris-courcelles-69             11
sevres-wolfenbuttel-1           11

The most popular station overall is Paris-Portedemontrouge-8 station

#### 1.63 b) Finding out the most popular station at the most picking hour

In [84]:
#We will use the most popular picking hour for blue cars

sp = df2.groupby('HOUR')['BLUECAR COUNTER'].sum()
sp.head()

HOUR
0    483
1    493
2    559
3    549
4    512
Name: BLUECAR COUNTER, dtype: int64

In [85]:
spp = df2.groupby(['HOUR', 'ID'])[['BLUECAR COUNTER']].sum()
spp.head()

BLUECAR COUNTER
HOUR ID                                                
0    alfortville-charlesdegaulle-16                   2
     alfortville-europe-22                            0
     antony-velpeau-14                                1
     arcueil-paulbert-21                              3
     asnieressurseine-henripoincare-53                0

In [87]:
sppp = spp.sort_values('BLUECAR COUNTER', ascending = False)
sppp


BLUECAR COUNTER
HOUR ID                                               
8    paris-portedevilliers-28                       15
2    courbevoie-paix-2                              15
     saintouen-louisblanc-35                        14
13   velizyvillacoublay-pauldautier-2               14
4    paris-claudebernard-16                         14
...                                                ...
17   paris-francois1er-38                            0
     paris-four-38                                   0
8    paris-legendre-126                              0
     paris-legendre-22                               0
2    paris-dante-1                                   0

[4612 rows x 1 columns]

The most popular station in this case is the Paris-portedevilliers-28 station

### 1.64) Finding out the most popular postal code for picking up blue cars

In [88]:
pc = df2.groupby(['HOUR', 'POSTAL CODE','ID'])[['POSTAL CODE']].count()
pc.head()

POSTAL CODE
HOUR POSTAL CODE ID                                       
0    75001       paris-croixdespetitschamps-4            1
                 paris-pyramides-12                      1
     75002       paris-alexandrie-7                      1
     75003       paris-bretagne-51                       1
                 paris-republique-17                     2

In [90]:
#we will rename the postal code column

pc.columns = ['POSTAL CODE COUNT']
pc.head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                             
0    75001       paris-croixdespetitschamps-4                  1
                 paris-pyramides-12                            1
     75002       paris-alexandrie-7                            1
     75003       paris-bretagne-51                             1
                 paris-republique-17                           2

In [91]:
ppc = pc.sort_values('POSTAL CODE COUNT', ascending = False)
ppc.head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                       
20   75010       paris-paradis-28                        4
21   75008       paris-tronchet-19                       4
     75011       paris-voltaire-182                      3
4    75005       paris-claudebernard-16                  3
5    75010       paris-chateaulandon-24                  3

In [92]:
pppc = ppc.groupby('POSTAL CODE')[['POSTAL CODE COUNT']].sum()
pppc.head()

POSTAL CODE COUNT
POSTAL CODE                   
75001                       55
75002                       61
75003                       67
75004                       68
75005                       90

In [93]:
#Let's sort the values to get the most popular postal code
ppc.sort_values('POSTAL CODE COUNT', ascending = False).head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                                        
20   75010       paris-paradis-28                                         4
21   75008       paris-tronchet-19                                        4
     75011       paris-voltaire-182                                       3
18   92130       issylesmoulineaux-promenadeduverger-166                  3
4    75005       paris-claudebernard-16                                   3

Therefore, The most popular postal codes in this case are 75010 and 75008






### 1.65) Finding out the stations at most popular postal code 




In [95]:
sppc = ppc.groupby(['POSTAL CODE','ID'])[['POSTAL CODE COUNT']].sum()
sppc.head()

POSTAL CODE COUNT
POSTAL CODE ID                                             
75001       paris-amiralcoligny-2                         3
            paris-bertinpoiree-14                         6
            paris-coloneldriant-7                         4
            paris-croixdespetitschamps-4                  6
            paris-halles-3                                4

In [96]:
#Let's sort these values 

sppc.sort_values('POSTAL CODE COUNT', ascending = False)

POSTAL CODE COUNT
POSTAL CODE ID                                           
75014       paris-portedemontrouge-8                   13
92310       sevres-wolfenbuttel-1                      11
75011       paris-voltaire-182                         11
75019       paris-mathis-35                            11
75008       paris-courcelles-69                        11
...                                                   ...
92270       boiscolombes-jeanmermoz-10                  1
75007       paris-cognacqjay-20                         1
94000       creteil-marechallyautey-16                  1
92320       chatillon-paris-21                          1
75014       paris-vercingetorix-106                     1

[1105 rows x 1 columns]

The station associated with the most popular code is Paris-portedemontrouge-8 at postal code 75014

### 1.66) Finding out the most popular postal code during the most picking hour

In [97]:
pppp = ppc.sort_values('POSTAL CODE COUNT', ascending = False)
pppp.head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                                        
20   75010       paris-paradis-28                                         4
21   75008       paris-tronchet-19                                        4
     75011       paris-voltaire-182                                       3
18   92130       issylesmoulineaux-promenadeduverger-166                  3
4    75005       paris-claudebernard-16                                   3

In [98]:
ppppp = ppc.loc[[21]]
pppp.head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                                        
20   75010       paris-paradis-28                                         4
21   75008       paris-tronchet-19                                        4
     75011       paris-voltaire-182                                       3
18   92130       issylesmoulineaux-promenadeduverger-166                  3
4    75005       paris-claudebernard-16                                   3

In [99]:
pppp.sort_values('POSTAL CODE COUNT',ascending = False)
pppp.head()

POSTAL CODE COUNT
HOUR POSTAL CODE ID                                                        
20   75010       paris-paradis-28                                         4
21   75008       paris-tronchet-19                                        4
     75011       paris-voltaire-182                                       3
18   92130       issylesmoulineaux-promenadeduverger-166                  3
4    75005       paris-claudebernard-16                                   3

The most popular postal codes during peak hours are 75008 & 75011

##1.7 Recommendation